From 5f98634037b682138be08fcd54fba909d3bff68c Mon Sep 17 00:00:00 2001
From: Peter Jones <pjones@redhat.com>
Date: Mon, 27 Feb 2017 10:21:00 -0500
Subject: [PATCH] Fix a definition gcc 7 dislikes

Signed-off-by: Peter Jones <pjones@redhat.com>
---
 linux/ucs2.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/linux/ucs2.h b/linux/ucs2.h
index 4ccfd28..1853453 100644
--- a/linux/ucs2.h
+++ b/linux/ucs2.h
@@ -10,7 +10,7 @@
  */
 static inline char *
 __attribute__((__unused__))
-ucs2_to_utf8(const uint16_t const *chars, size_t max)
+ucs2_to_utf8(const uint16_t *chars, size_t max)
 {
 	size_t i, j;
 	char *ret = alloca(max * 6 + 1);
-- 
2.13.0

